### DISCUSSION TITLE
[Proposal] 'notify' keyword for properties that need to invoke INotifyPropertyChanged.

### DISCUSSION THREAD

**Discussion by Gouellie** [dotnet/csharplang#5026] [Language Ideas] [answered]
A reserve keyword for properties that invoke a basic implementation of INotifyPropertyChanged. 
The idea is to reduce amount of boilerplate code and the footprint of simple properties, especially when used in the context of a WPF application using a MVVM pattern.

Example

```csharp

// A property that requires a backing field in order to invoke RaisePropertyChanged() in its setter.

        private bool _isSelected;
        public bool IsSelected
        {
            get => _isSelected;
            set
            {
                _isSelected = value;
                RaisePropertyChanged();
            }
        }

// Could now be written as such.
        public bool notify IsSelected { get ; set; }

```

**Comment by bernd5** [comment]
For things like that you could use the generator approach. (is there already a generator which implements this?)
```cs
[Notify]
private bool _isSelected;
```

**Reply by ufcpp** [reply]
https://github.com/dotnet/roslyn-sdk/blob/main/samples/CSharp/SourceGenerators/SourceGeneratorSamples/AutoNotifyGenerator.cs

**Reply by Gouellie** [reply]
That's an interesting solution. 👍

**Comment by jnm2** [comment, marked as answer]
One alternative way to reduce boilerplate is https://github.com/dotnet/csharplang/issues/140.

Today, I write these properties using a base class that defines a `Set` method that handles equality comparison and notification:

```cs
private bool isSelected;
public bool IsSelected { get => isSelected; set => Set(ref isSelected, value); }
```

When https://github.com/dotnet/csharplang/issues/140 ships, I will be able to write:

```cs
public bool IsSelected { get; set => Set(ref field, value); }
```

My base class:

```cs
using System.ComponentModel;
using System.Runtime.CompilerServices;

public abstract class ObservableObject : INotifyPropertyChanged
{
    public event PropertyChangedEventHandler? PropertyChanged;

    protected virtual void OnPropertyChanged([CallerMemberName] string? propertyName = null)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }

    protected bool Set<T>(
        [NotNullIfNotNull(nameof(value))] ref T location,
        T value,
        [CallerMemberName] string? propertyName = null)
    {
        var shouldDetectChange = !EqualityComparer<T>.Default.Equals(location, value);
        location = value;
        if (!shouldDetectChange) return false;
        OnPropertyChanged(propertyName);
        return true;
    }
}
```

**Reply by Gouellie** [reply]
Yes! I'm also using that trick, I didn't know about the field keyword proposal, I agree that once it ships this should be a satisfactory solution.
